pkg/upnp/upnp.go

```diff
@@ -0,0 +1,127 @@
+package upnp
+
+import (
+	"context"
+	"strings"
+
+	"github.com/huin/goupnp"
+	"github.com/huin/goupnp/dcps/internetgateway2"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// The UPNP library is structured a bit funny. To create a forward on a gateway and know
+// the external IP of that gateway we need two different instances of the UPNP client
+// WANIPv6FirewallControlClient can be empty if the gatway doesn't support the WANIPv6FirewallControl service
+type Gateway struct {
+	ConnectionClient             ConnectionClient
+	WANIPv6FirewallControlClient *internetgateway2.WANIPv6FirewallControl1
+}
+
+type ConnectionClient interface {
+	AddPortMapping(
+		NewRemoteHost string,
+		NewExternalPort uint16,
+		NewProtocol string,
+		NewInternalPort uint16,
+		NewInternalClient string,
+		NewEnabled bool,
+		NewPortMappingDescription string,
+		NewLeaseDuration uint32,
+	) (err error)
+
+	GetExternalIPAddress() (
+		NewExternalIPAddress string,
+		err error,
+	)
+	GetServiceClient() *goupnp.ServiceClient
+}
+
+func GetGatewayClients(ctx context.Context) []Gateway {
+	clients := GetConnectionClients(ctx)
+	gatewayClients := make([]Gateway, len(clients))
+
+	for i := range clients {
+		gatewayClients[i] = Gateway{ConnectionClient: clients[i], WANIPv6FirewallControlClient: nil}
+		gatewayURL := clients[i].GetServiceClient().Location
+		if wanipv6clients, err := internetgateway2.NewWANIPv6FirewallControl1ClientsByURLCtx(ctx, gatewayURL); err == nil {
+			gatewayClients[i].WANIPv6FirewallControlClient = wanipv6clients[0]
+		} else {
+			log.Warnf("Unable to find WANIPv6FirewallControl1Clients for Gateway %s [%s]", gatewayURL, err.Error())
+		}
+	}
+	return gatewayClients
+}
+
+// Gather UPNPConnectionClients in the network and treats errors as non-critical. Use this to find out the external IPs of the network and configure port forwarding
+func GetConnectionClients(ctx context.Context) []ConnectionClient {
+	tasks, _ := errgroup.WithContext(ctx)
+	// This code sucks and I did not manage to make it more concise.
```

Comment:
haha, love this comment

Reply:
Yeah, had a lot of fun with the UPNP library. It is really not fun to work with half a dozen auto generated SOAP clients. Unfortunately this seems to be the intended way to do this: https://github.com/huin/goupnp/blob/main/GUIDE.md#internet-gateways

Are you ok with this implementation?